akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/AbstractStashSpec.scala

```diff
@@ -0,0 +1,599 @@
+    }
+
+    "signal PostStop to the latest unstashed behavior on failure with stop supervision" in {
+      val probe = TestProbe[String]()
+      val ref =
+        spawn(Behaviors.supervise(stashingBehavior(probe.ref))
+          .onFailure[TestException](SupervisorStrategy.stop))
+      testPostStop(probe, ref)
+    }
+
+    def testPreRestart(
+      probe:      TestProbe[String],
+      childLatch: Option[CountDownLatch],
+      ref:        ActorRef[String]
+    ): Unit = {
+      ref ! "stash"
+      ref ! "stash"
+      ref ! "stash-fail"
+      ref ! "stash"
+      EventFilter[TestException](
+        start = "Supervisor RestartSupervisor saw failure: unstash-fail",
+        occurrences = 1
+      ).intercept {
+        ref ! "unstash"
+        // when childLatch is defined this be stashed in the internal stash of the RestartSupervisor
+        // because it's waiting for child to stop
+        ref ! "get-current"
+
+        probe.expectMessage("unstashing-0")
+        probe.expectMessage("unstashing-1")
+        probe.expectMessage("stash-fail-2")
+        probe.expectMessage("pre-restart-2")
+
+        childLatch.foreach(_.countDown())
+        probe.expectMessage("current-00")
+
+        ref ! "get-stash-size"
+        probe.expectMessage("stash-size-0")
+      }
+    }
+
+    "signal PreRestart to the latest unstashed behavior on failure with restart supervision" in {
+      val probe = TestProbe[String]()
+      val ref =
+        spawn(Behaviors.supervise(stashingBehavior(probe.ref))
+          .onFailure[TestException](SupervisorStrategy.restart))
+
+      testPreRestart(probe, None, ref)
+      // one more time to ensure that the restart strategy is kept
+      testPreRestart(probe, None, ref)
+
+      Thread.sleep(2000)
+      println(s"# DONE -----------") // FIXME
+    }
+
+    "signal PreRestart to the latest unstashed behavior on failure with restart supervision and slow stopping child" in {
+      val probe = TestProbe[String]()
+      val childLatch = new CountDownLatch(1)
+      val ref =
+        spawn(Behaviors.supervise(stashingBehavior(probe.ref, Some(childLatch)))
+          .onFailure[TestException](SupervisorStrategy.restart))
+
+      testPreRestart(probe, Some(childLatch), ref)
+
+      Thread.sleep(2000)
+      println(s"# DONE -----------") // FIXME
+    }
+
+    "signal PreRestart to the latest unstashed behavior on failure with backoff supervision" in {
+      val probe = TestProbe[String]()
+      val ref =
+        spawn(Behaviors.supervise(stashingBehavior(probe.ref))
+          .onFailure[TestException](SupervisorStrategy.restartWithBackoff(100.millis, 100.millis, 0.0)))
+
+      testPreRestart(probe, None, ref)
+
+      // FIXME 26148 Note ClassCastException: akka.actor.typed.internal.RestartSupervisor$ResetRestartCount cannot
+      // be cast to java.lang.String
+      // which is because the original backoff strategy isn't installed after the RestartSupervisor has unstashAll
```

Comment:
This is where I stopped. Supervision is also using the stash while restarting (waiting for children) or when backoff.